types/handler.go

```diff
@@ -5,4 +5,36 @@ type Handler func(ctx Context, msg Msg) Result
 
 // AnteHandler authenticates transactions, before their internal messages are handled.
 // If newCtx.IsZero(), ctx is used instead.
-type AnteHandler func(ctx Context, tx Tx, simulate bool) (newCtx Context, result Result, abort bool)
+type AnteHandler func(ctx Context, tx Tx, simulate bool) (newCtx Context, err error)
+
+// AnteDecorator wraps the next AnteHandler to perform custom pre- and post-processing.
+type AnteDecorator interface {
+	AnteHandle(ctx Context, tx Tx, simulate bool, next AnteHandler) (newCtx Context, err error)
+}
+
+// ChainDecorator chains AnteDecorators together with each element
+// wrapping over the decorators further along chain and returns a single AnteHandler.
+//
+// First element is outermost decorator, last element is innermost decorator
+func ChainAnteDecorators(chain ...AnteDecorator) AnteHandler {
+	if (chain[len(chain)-1] != Terminator{}) {
```

Comment:
Is there a reason for the extra () around this?

Reply:
Compiler gets confused with ending characters: `{} {`

Reply:
Yes, I agree. But keep in mind, this will not play nicely or even work at all with `BaseApp`. The `BaseApp` assumes the `GasMeter` has been set correctly for each tx since it consumes gas for this. We need to document this very very clearly.